receiver/oktareceiver/logs.go

```diff
@@ -0,0 +1,236 @@
+	defer r.wg.Done()
+	t := time.NewTicker(r.cfg.PollInterval)
+
+	err := r.poll(ctx)
+	if err != nil {
+		r.logger.Error("there was an error during the first poll", zap.Error(err))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			err := r.poll(ctx)
+			if err != nil {
+				r.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *oktaLogsReceiver) poll(ctx context.Context) error {
+	logEvents := r.getLogs(ctx)
+	observedTime := pcommon.NewTimestampFromTime(time.Now())
+	logs := r.processLogEvents(observedTime, logEvents)
+	if logs.LogRecordCount() > 0 {
+		if err := r.consumer.ConsumeLogs(ctx, logs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *oktaLogsReceiver) getLogs(ctx context.Context) []*okta.LogEvent {
+	var logs []*okta.LogEvent
+	var reqURL string
+	pollTime := time.Now().UTC()
+
+	// get logs until there isn't any overflow OR we get logs published after initial pollTime
+	for {
+		if r.nextURL == "" {
+			reqURL = "https://" + r.cfg.Domain + "/api/v1/logs"
+		} else {
+			reqURL = r.nextURL
+		}
+
+		req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
+		if err != nil {
+			r.logger.Error("error creating okta api request", zap.Error(err))
+			break
+		}
+
+		// add query params to the first polling request
+		if r.nextURL == "" {
+			query := req.URL.Query()
+			query.Add("since", pollTime.Add(-r.cfg.PollInterval).Format(OktaTimeFormat))
+			query.Add("limit", strconv.Itoa(oktaMaxLimit))
+			req.URL.RawQuery = query.Encode()
+		}
+
+		req.Header.Add("Authorization", "SSWS "+string(r.cfg.APIToken))
+
+		res, err := r.client.Do(req)
+		if err != nil {
+			r.logger.Error("error performing okta api request", zap.Error(err))
+			break
+		}
+
+		defer res.Body.Close()
+
+		if res.StatusCode != http.StatusOK {
+			r.logger.Error("okta logs endpoint returned non-200 statuscode: " + res.Status)
+			break
+		}
+
+		r.setNextLink(res)
+
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			r.logger.Error("error reading response body", zap.Error(err))
+			break
+		}
+
+		var curLogs []*okta.LogEvent
+		err = json.Unmarshal(body, &curLogs)
+		if err != nil {
+			r.logger.Error("unable to unmarshal log events", zap.Error(err))
+			break
+		}
+
+		logs = append(logs, curLogs...)
+
+		if len(curLogs) < oktaMaxLimit || curLogs[len(curLogs)-1].Published.After(pollTime) {
+			break
+		}
+	}
+
+	return logs
+}
+
+func (r *oktaLogsReceiver) processLogEvents(observedTime pcommon.Timestamp, logEvents []*okta.LogEvent) plog.Logs {
+	logs := plog.NewLogs()
+
+	for _, logEvent := range logEvents {
+		resourceLogs := logs.ResourceLogs().AppendEmpty()
+		resourceLogs.ScopeLogs().AppendEmpty()
+		logRecord := resourceLogs.ScopeLogs().At(0).LogRecords().AppendEmpty()
+
+		// resource attributes
+		resourceAttributes := resourceLogs.Resource().Attributes()
+		resourceAttributes.PutStr("okta.domain", r.cfg.Domain)
```

Comment:
Let's pull this out of the loop, each log record should be nested under the same resource and scope, since they have the same resource/scope attributes for each.